xds/src/main/java/com/linecorp/armeria/xds/client/endpoint/XdsHealthCheckedEndpointGroupBuilder.java

```diff
@@ -0,0 +1,103 @@
+import com.linecorp.armeria.internal.client.endpoint.healthcheck.HttpHealthChecker;
+
+import io.envoyproxy.envoy.config.cluster.v3.Cluster;
+import io.envoyproxy.envoy.config.core.v3.HealthCheck.HttpHealthCheck;
+import io.envoyproxy.envoy.config.endpoint.v3.Endpoint.HealthCheckConfig;
+import io.envoyproxy.envoy.config.endpoint.v3.LbEndpoint;
+
+final class XdsHealthCheckedEndpointGroupBuilder
+        extends AbstractHealthCheckedEndpointGroupBuilder<XdsHealthCheckedEndpointGroupBuilder> {
+
+    private static final Logger logger = LoggerFactory.getLogger(XdsHealthCheckedEndpointGroupBuilder.class);
+
+    private final Cluster cluster;
+    private final HttpHealthCheck httpHealthCheck;
+
+    XdsHealthCheckedEndpointGroupBuilder(EndpointGroup delegate, Cluster cluster,
+                                         HttpHealthCheck httpHealthCheck) {
+        super(delegate);
+        this.cluster = cluster;
+        this.httpHealthCheck = httpHealthCheck;
+    }
+
+    @Override
+    protected Function<? super HealthCheckerContext, ? extends AsyncCloseable> newCheckerFactory() {
+        return ctx -> {
+            final LbEndpoint lbEndpoint = EndpointUtil.lbEndpoint(ctx.originalEndpoint());
+            final HealthCheckConfig healthCheckConfig = lbEndpoint.getEndpoint().getHealthCheckConfig();
+            final String path = httpHealthCheck.getPath();
+            final String host = Strings.emptyToNull(httpHealthCheck.getHost());
+
+            final HttpHealthChecker checker =
+                    new HttpHealthChecker(ctx, endpoint(healthCheckConfig, ctx.originalEndpoint()),
+                                          path, httpMethod(httpHealthCheck) == HttpMethod.GET,
+                                          protocol(cluster), host);
+            checker.start();
+            return checker;
+        };
+    }
+
+    private static HttpMethod httpMethod(HttpHealthCheck httpHealthCheck) {
+        final HttpMethod method = EndpointUtil.convert(httpHealthCheck.getMethod(), HttpMethod.GET);
+        if (method != HttpMethod.GET && method != HttpMethod.HEAD) {
+            logger.warn("Unsupported http method: {}. Only GET and HEAD are supported. " +
+                        "Falling back to GET for health checks.", method);
+            return HttpMethod.GET;
+        }
+        return method;
+    }
+
+    private static SessionProtocol protocol(Cluster cluster) {
+        // Not using httpHealthCheck.getCodecClientType() because
+        // HTTP[S] covers both HTTP/1 and HTTP/2.
+        if (EndpointUtil.isTls(cluster)) {
+            return SessionProtocol.HTTPS;
+        } else {
+            return SessionProtocol.HTTP;
+        }
+    }
+
+    private static Endpoint endpoint(HealthCheckConfig healthCheckConfig, Endpoint endpoint) {
+        if (healthCheckConfig == HealthCheckConfig.getDefaultInstance()) {
+            return endpoint;
+        }
+        final int port = healthCheckConfig.getPortValue();
```

Comment:
`port` is optional so it could be `0`. If so, an exception will be raised in `validatePort`.